File: src/codeforces/contest1760/ProblemF.java
package codeforces.contest1760;

import java.util.Arrays;
import java.util.Scanner;

public class ProblemF {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int tests = scanner.nextInt();
        for (int t = 0; t < tests; t++) {
            int n = scanner.nextInt();
            long c = scanner.nextLong();
            int d = scanner.nextInt();
            int[] quests = new int[n];
            for (int i = 0; i < n; i++) quests[i] = scanner.nextInt();
            int ans = solve(quests, d, c);
            if (ans < 0) System.out.println("Impossible");
            else if (ans > d) System.out.println("Infinity");
            else System.out.println(ans);
        }
    }

    private static int solve(int[] quests, int d, long c) {
        int n = quests.length;
        Arrays.sort(quests);
        if ((long) quests[n - 1] * d < c) return -1;
        long[] prefSum = new long[n + 1];
        for (int i = 0, j = n - 1; i < n; i++, j--) prefSum[i + 1] = prefSum[i] + quests[j];
        if (n <= d && prefSum[n] >= c || n >= d && prefSum[d] >= c) return d + 1;

        int lo = 0, hi = d - 2;
        while (lo < hi) {
            int mid = lo + (hi - lo + 1) / 2;
            if (getVal(prefSum, n, d, mid) >= c) lo = mid;
            else hi = mid - 1;
        }
        return lo;
    }

    private static long getVal(long[] pref, int n, int d, int k) {
        return pref[Math.min(n, k + 1)] * (d / (k + 1)) + pref[Math.min(n, d % (k + 1))];
    }
}

File: src/codeforces/contest1862/ProblemF.kt
@file:Suppress("DuplicatedCode")

package codeforces.contest1862

import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max
import kotlin.math.min

fun main(args: Array<String>) {
    val tests = nextInt()
    for (t in 0 until tests) {
        val w = nextInt()
        val f = nextInt()
        val n = nextInt()
        val s = nextInts(n)
        println(solve(f, w, s))
    }
}

fun solve(w: Int, f: Int, s: IntArray): Int {
    val total = s.sum()
    val bitset = BitSet(total + 1)
    bitset.set(0)
    var curSum = 0
    for (num in s) {
        var tmp = curSum
        while (tmp >= 0) {
            bitset.set(tmp + num)
            tmp = bitset.previousSetBit(tmp - 1)
        }
        curSum += num
    }
    var ans = total
    var tmp = 0
    while (tmp != -1) {
        ans = min(ans, max((tmp + f - 1) / f, (total - tmp + w - 1) / w))
        tmp = bitset.nextSetBit(tmp + 1)
    }
    return ans
}

private class MyReader1(inputStream: InputStream) {
    private val reader = BufferedReader(InputStreamReader(inputStream))
    private var tokenizer: StringTokenizer? = null

    fun next(): String {
        while (true) {
            tokenizer.let {
                if (it == null || !it.hasMoreTokens()) {
                    tokenizer = StringTokenizer(reader.readLine())
                } else {
                    return it.nextToken()
                }
            }
        }
    }
}

private val reader = MyReader1(System.`in`)
private fun next() = reader.next()
private fun nextInt() = next().toInt()
private fun nextLong() = next().toLong()
private fun nextDouble() = next().toDouble()
private fun nextStrings(n: Int) = Array(n) { next() }
private fun nextInts(n: Int) = IntArray(n) { nextInt() }
private fun nextLongs(n: Int) = LongArray(n) { nextLong() }
private fun nextDoubles(n: Int) = DoubleArray(n) { nextDouble() }